libvast/test/module.cpp

```diff
@@ -0,0 +1,689 @@
+    if (!type_or_error)
+      return type_or_error.error();
+    record_fields.emplace_back(record_record.begin()->first, *type_or_error);
+  }
+  return record_type{record_fields};
+}
+
+type get_known_type(const std::vector<type>& known_types,
+                    std::string_view name) {
+  for (const auto& known_type : known_types) {
+    if (name == known_type.name()) {
+      return known_type;
+    }
+  }
+  return {}; // none type
+}
+
+std::vector<std::string> reserved_names
+  = {"bool",    "integer", "count",  "real", "duration", "time", "string",
+     "pattern", "address", "subnet", "enum", "list",     "map",  "record"};
+
+caf::expected<type> to_type(const std::vector<type>& known_types,
+                            const data& declaration, std::string_view name) {
+  const auto* known_type_name_ptr = caf::get_if<std::string>(&declaration);
+  // Prevent using reserved names as types names
+  if (std::any_of(reserved_names.begin(), reserved_names.end(),
+                  [&](const auto& reserved_name) {
+                    return name == reserved_name;
+                  }))
+    return caf::make_error(
+      ec::parse_error,
+      fmt::format("parses a new type with a reserved name: {}", name));
+  // Type names can contain any character that the YAML parser can handle - no
+  // need to check for allowed characters.
+  if (known_type_name_ptr != nullptr) {
+    const auto& known_type_name = *known_type_name_ptr;
+    // Check built-in types first
+    if (known_type_name == "bool")
+      return type{name, bool_type{}};
+    if (known_type_name == "integer")
+      return type{name, integer_type{}};
+    if (known_type_name == "count")
+      return type{name, count_type{}};
+    if (known_type_name == "real")
+      return type{name, real_type{}};
+    if (known_type_name == "duration")
+      return type{name, duration_type{}};
+    if (known_type_name == "time")
+      return type{name, time_type{}};
+    if (known_type_name == "string")
+      return type{name, string_type{}};
+    if (known_type_name == "pattern")
+      return type{name, pattern_type{}};
+    if (known_type_name == "address")
+      return type{name, address_type{}};
+    if (known_type_name == "subnet")
+      return type{name, subnet_type{}};
+    // Check type aliases aka. known types
+    const auto& known_type = get_known_type(known_types, known_type_name);
+    if (!known_type)
+      return caf::make_error(ec::parse_error, fmt::format("found unknown type: "
+                                                          "{}",
+                                                          known_type_name));
+    return type{name, known_type};
+  }
+  const auto* decl_ptr = caf::get_if<record>(&declaration);
+  if (decl_ptr == nullptr)
+    return caf::make_error(ec::parse_error, "parses type alias with invalid "
+                                            "format");
+  const auto& decl = *decl_ptr;
```

Comment:
Naming this `decl` with an argument named `declaration` is rather confusing.

Please use consistent names throughout the function so it makes more sense. Maybe it's `declaration_record` or so.